design/src/rfcs/rfc0017_customizable_client_operations.md

```diff
@@ -0,0 +1,185 @@
+similar to `send`, but that returns a customizable request. The customer experience should look as
+follows:
+
+```rust
+let response = client.some_operation()
+    .some_value(5)
+    .customize()?
+    .map_request(|mut req| {
+        req.headers_mut().insert(
+            HeaderName::from_static("x-some-header"),
+            HeaderValue::from_static("some-value")
+        );
+        Result::<_, Infallible>::Ok(req)
+    })?
+    .send()
+    .await?;
+```
+
+This new `customize` method would return the following:
+
+```rust
+pub struct CustomizableOperation<O, R> {
+    handle: Arc<Handle>,
+    operation: Operation<O, R>,
+}
+
+impl<O, R> CustomizableOperation<O, R> {
+    // Allows for customizing the operation's request
+    fn map_request<E>(
+        mut self,
+        f: impl FnOnce(Request<SdkBody>) -> Result<Request<SdkBody>, E>,
+    ) -> Result<Self, E> {
+        let (request, response) = self.operation.into_request_response();
+        let request = request.augment(|req, _props| f(req))?;
+        self.operation = Operation::from_parts(request, response);
+        Ok(self)
+    }
+
+    // Allows for customizing the entire operation
+    fn map_operation<E>(
+        mut self,
+        f: impl FnOnce(Operation<O, R>) -> Result<Operation<O, R>, E>,
+    ) -> Result<Self, E> {
+        self.operation = f(self.operation)?;
+        Ok(self)
+    }
+
+    // Direct access to read the request
+    fn request(&self) -> &Request<SdkBody> {
+        self.operation.request()
+    }
+
+    // Direct access to mutate the request
+    fn request_mut(&mut self) -> &mut Request<SdkBody> {
+        self.operation.request_mut()
+    }
+
+    // Sends the operation's request
+    async fn send<T, E>(self) -> Result<T, SdkError<E>>
+    where
+        O: ParseHttpResponse<Output = Result<T, E>> + Send + Sync + Clone + 'static,
+        E: std::error::Error,
+        R: ClassifyResponse<SdkSuccess<T>, SdkError<E>> + Send + Sync,
+    {
+        self.handle.client.call(self.operation).await
+    }
+}
+```
+
+Additionally, for those who want to avoid closures, the `Operation` type will have
+`request` and `request_mut` methods added to it to get direct access to its underlying
+HTTP request.
+
+The `CustomizableOperation` type will then mirror these functions so that the experience
+can look as follows:
+
+```rust
+let mut operation = client.some_operation()
+    .some_value(5)
+    .customize()?;
+operation.request_mut()
+    .headers_mut()
+    .insert(
+        HeaderName::from_static("x-some-header"),
+        HeaderValue::from_static("some-value")
+    );
+let response = operation.send().await?;
+```
+
+### Eliminating `async` from `make_operation`
+
+To avoid requiring an `await` after the call to `customize`, the `make_operation` function
+generated on input structs needs to be made synchronous. The `make_operation` function
+originally was synchronous, but was made async during the implementation of the Glacier
```

Comment:
I think I'd probaby keep it async and require an await after customize. I think keeping make_operation async is fairly important from the perspective of keeping the design forwards compatible